Classes/AbstractDataHandlerListener.php

```diff
@@ -138,6 +171,142 @@ protected function getAllChangeSetValuesMatch($triggerConfiguration, $changedFie
+                break;
+            }
+        }
+
+        return $name;
+    }
+
+    /**
+     * Retrieves a record, taking into account the additionalWhereClauses of the
+     * Indexing Queue configurations.
+     *
+     * @param string $recordTable Table to read from
+     * @param int $recordUid Id of the record
+     * @param TypoScriptConfiguration $solrConfiguration
+     * @return array Record if found, otherwise empty array
+     */
+    protected function getRecord($recordTable, $recordUid, TypoScriptConfiguration $solrConfiguration)
+    {
+        $record = [];
+        $indexingConfigurations = $solrConfiguration->getEnabledIndexQueueConfigurationNames();
+
+        foreach ($indexingConfigurations as $indexingConfigurationName) {
+            $record = $this->getRecordWhenIndexConfigurationIsValid($recordTable, $recordUid,
+                $indexingConfigurationName, $solrConfiguration);
+            if (!empty($record)) {
+                // if we found a record which matches the conditions, we can continue
+                break;
+            }
+        }
+
+        return $record;
+    }
+
+    /**
+     * This method return the record array if the table is valid for this indexingConfiguration.
+     * Otherwise an empty array will be returned.
+     *
+     * @param string $recordTable
+     * @param integer $recordUid
+     * @param string $indexingConfigurationName
+     * @param TypoScriptConfiguration $solrConfiguration
+     * @return array
+     */
+    private function getRecordWhenIndexConfigurationIsValid(
+        $recordTable,
+        $recordUid,
+        $indexingConfigurationName,
+        TypoScriptConfiguration $solrConfiguration
+    ) {
+        if (!$this->isValidTableForIndexConfigurationName($recordTable, $indexingConfigurationName,
+            $solrConfiguration)
+        ) {
+            return [];
+        }
+
+        $recordWhereClause = $solrConfiguration->getIndexQueueAdditionalWhereClauseByConfigurationName($indexingConfigurationName);
+
+        if ($recordTable === 'pages_language_overlay') {
+            return $this->getPageOverlayRecordWhenParentIsAccessible($recordUid, $recordWhereClause);
+        }
+
+        return (array)BackendUtility::getRecord($recordTable, $recordUid, '*', $recordWhereClause);
+    }
+
+    /**
+     * This method is used to check if a table is an allowed table for an index configuration.
+     *
+     * @param string $recordTable
+     * @param string $indexingConfigurationName
+     * @param TypoScriptConfiguration $solrConfiguration
+     * @return boolean
+     */
+    private function isValidTableForIndexConfigurationName(
+        $recordTable,
+        $indexingConfigurationName,
+        TypoScriptConfiguration $solrConfiguration
+    ) {
+        $tableToIndex = $solrConfiguration->getIndexQueueTableNameOrFallbackToConfigurationName($indexingConfigurationName);
+
+        $isMatchingTable = $tableToIndex === $recordTable;
```

Comment:
when you have boolean evaluations, it's better for readability to surround them with parentheses: `$isMatchingTable = ($tableToIndex === $recordTable);`